### PR TITLE
Implemented simple representation of ArrayList

### PR DESCRIPTION
**Pull request by Likh-Alex**

### CODE REVIEW

File: src/main/java/core/basesyntax/ArrayList.java [hunk 1]
```diff
@@ -1,49 +1,106 @@
...
+    }
 
     @Override
     public void add(T value) {
-
+        if (size == elements.length) {
+            resize();
+        }
+        elements[size] = value;
+        size++;
```
Comment (dmytro-hryhoruk) [inline src/main/java/core/basesyntax/ArrayList.java:27]:
```suggestion
        elements[size++] = value;
```

File: src/main/java/core/basesyntax/ArrayList.java [hunk 2]
```diff
@@ -1,49 +1,106 @@
...
+            resize();
+        } else {
+            System.arraycopy(elements, index, elements, index + 1, size - index);
+            if (index == size) {
+                add(value);
+            } else {
+                set(value, index);
+                size++;
+            }
+        }
```
Comment (dmytro-hryhoruk) [inline src/main/java/core/basesyntax/ArrayList.java:45]:
```suggestion
        }
```

File: src/main/java/core/basesyntax/ArrayList.java [hunk 3]
```diff
@@ -1,49 +1,106 @@
...
-
+        for (int i = 0; i < list.size(); i++) {
+            add(list.get(i));
+        }
     }
 
     @Override
     public T get(int index) {
-        return null;
+        if (index >= size || index < 0) {
```
Comment (dmytro-hryhoruk) [inline src/main/java/core/basesyntax/ArrayList.java:57]: let's make a boolean method and use it in get(), set(), and remove()

File: src/main/java/core/basesyntax/ArrayList.java [hunk 4]
```diff
@@ -1,49 +1,106 @@
...
+            throw new ArrayIndexOutOfBoundsException(index + " is out of bounds");
+        } else {
```
Comment (dmytro-hryhoruk) [inline src/main/java/core/basesyntax/ArrayList.java:59]: you don't need else after throw new Exception

File: src/main/java/core/basesyntax/ArrayList.java [hunk 5]
```diff
@@ -1,49 +1,106 @@
...
+            return elements[index];
+        }
     }
 
     @Override
     public void set(T value, int index) {
-
+        if (index >= size || index < 0) {
+            throw new ArrayIndexOutOfBoundsException(index + " is out of bounds");
+        } else {
```
Comment (dmytro-hryhoruk) [inline src/main/java/core/basesyntax/ArrayList.java:68]: same

File: src/main/java/core/basesyntax/ArrayList.java [hunk 6]
```diff
@@ -1,49 +1,106 @@
...
+            elements[index] = value;
+        }
     }
 
     @Override
     public T remove(int index) {
-        return null;
+        if (index >= size || index < 0) {
+            throw new ArrayIndexOutOfBoundsException(index + " is out of bounds");
+        } else {
```
Comment (dmytro-hryhoruk) [inline src/main/java/core/basesyntax/ArrayList.java:77]: and here

File: src/main/java/core/basesyntax/ArrayList.java [hunk 7]
```diff
@@ -1,49 +1,106 @@
...
     }
 
     @Override
     public T remove(T t) {
-        return null;
+        for (int i = 0; i < size(); i++) {
+            if (elements[i] == t || elements[i] != null && elements[i].equals(t)) {
+                System.arraycopy(elements,i + 1, elements, i,size - i);
+                size--;
+                return t;
```
Comment (dmytro-hryhoruk) [inline src/main/java/core/basesyntax/ArrayList.java:91]:
```suggestion
                return remove(i);
```

File: src/main/java/core/basesyntax/ArrayList.java [hunk 8]
```diff
@@ -1,49 +1,106 @@
...
+            resize();
+        } else {
+            System.arraycopy(elements, index, elements, index + 1, size - index);
+            if (index == size) {
+                add(value);
+            } else {
+                set(value, index);
+                size++;
+            }
+        }
```
Comment (dmytro-hryhoruk) [inline src/main/java/core/basesyntax/ArrayList.java:44]: to many if-else blocks. Try to simplify that. Why you are first copying the array if there is a change of add() usage

**Review by dmytro-hryhoruk — CHANGES_REQUESTED** [review]

File: src/main/java/core/basesyntax/ArrayList.java (hunk 1 above)
Comment (Likh-Alex) [reply src/main/java/core/basesyntax/ArrayList.java:27]: done

File: src/main/java/core/basesyntax/ArrayList.java (hunk 2 above)
Comment (Likh-Alex) [reply src/main/java/core/basesyntax/ArrayList.java]: done

File: src/main/java/core/basesyntax/ArrayList.java (hunk 3 above)
Comment (Likh-Alex) [reply src/main/java/core/basesyntax/ArrayList.java]: Implemented method isInvalidIndex for get(), set() and remove()

File: src/main/java/core/basesyntax/ArrayList.java (hunk 4 above)
Comment (Likh-Alex) [reply src/main/java/core/basesyntax/ArrayList.java]: removed

File: src/main/java/core/basesyntax/ArrayList.java (hunk 5 above)
Comment (Likh-Alex) [reply src/main/java/core/basesyntax/ArrayList.java]: removed

File: src/main/java/core/basesyntax/ArrayList.java (hunk 6 above)
Comment (Likh-Alex) [reply src/main/java/core/basesyntax/ArrayList.java]: removed

File: src/main/java/core/basesyntax/ArrayList.java (hunk 7 above)
Comment (Likh-Alex) [reply src/main/java/core/basesyntax/ArrayList.java]: done

File: src/main/java/core/basesyntax/ArrayList.java (hunk 8 above)
Comment (Likh-Alex) [reply src/main/java/core/basesyntax/ArrayList.java]: tried to simplify as much as possible

File: src/main/java/core/basesyntax/ArrayList.java [hunk 9]
```diff
@@ -1,49 +1,99 @@
...
     @Override
     public int size() {
-        return 0;
+        return size;
     }
 
     @Override
     public boolean isEmpty() {
-        return false;
+        return size() == 0;
```
Comment (dmytro-hryhoruk) [inline src/main/java/core/basesyntax/ArrayList.java:97]:
```suggestion
        return size == 0;
```

**Review by dmytro-hryhoruk — APPROVED** [review]

File: src/main/java/core/basesyntax/ArrayList.java (hunk 9 above)
Comment (Likh-Alex) [reply src/main/java/core/basesyntax/ArrayList.java]: done as well

**Review by boroda4436 — APPROVED** [review]

**Pull request closed without merging**